### PR TITLE
169 | 4 MetaData and PreprocessCT refactored, grt123 processing adopted

### PR DESCRIPTION
**Pull request by vessemer**
Refactored classes `PreprocessCT` and `MetaData`. Preprocessing of grt123 algorithm has been replaced by them. 

## Description
`PreprocessCT` is now a subclass of `Params`, both are documented along with `MetaData`.
Correct me if I've mistaken, but there is a bug in the [processing part](https://github.com/concept-to-clinic/concept-to-clinic/commit/45db7be01d0ebcd4a91a071b4055eee4aac949cf#diff-af6ce34696667e4dea35064c3e319127R294) of the grt123 prediction algorithm adoptation:

```python
image = sitk.GetArrayFromImage(image_itk)
spacing = np.array(image_itk.GetSpacing())[::-1]  
image = lum_trans(image)
image = resample(image, spacing, np.array([1, 1, 1]), order=1)[0]
```  
- `spacing` will became redundant after re-sampling. 
```python
for nodule in nodule_list:
    print(nodule)
    nod_location = np.array([np.float32(nodule[s]) for s in ["z", "y", "x"]])
    nod_location *= spacing 
```  
- estimation of the `nod_location` should  also take into account both the `origin` of the CT scan and `new_spacing = [1, 1, 1]` and therefore shal looks like  
```python  
nod_location = (nod_location - origin) / new_spacing
```
Neglection of the `origin` and `new_spacing` leads to estimation of a complitely different CT part.

- the old estimation of patches by  [`SimpleCrop`](https://github.com/concept-to-clinic/concept-to-clinic/commit/45db7be01d0ebcd4a91a071b4055eee4aac949cf#diff-af6ce34696667e4dea35064c3e319127R226) requires `np.pad` to be applied on each patch independently for the same CT scan.  

The aforementioned bugs were corrected along with the adoptation of grt123 preprocessing via [`preprocess_ct.py`](https://github.com/vessemer/concept-to-clinic/blob/389824f6042a5cbc4cb537c531ae40446ab6164a/prediction/src/preprocess/preprocess_ct.py)

## Reference to official issue
This addresses #169 and partially covers #4 

## How Has This Been Tested?
For the `grt123_preprocess`, all tests lie in `test_grt123_preprocess` and are written in order to ensure fully interchangeability. 

## CLA
- [x] I have signed the CLA; if other committers are in the commit history, they have signed the CLA as well

### CODE REVIEW

File: prediction/src/algorithms/classify/src/gtr123_model.py [hunk 1]
```diff
@@ -288,23 +242,17 @@ def predict(image_itk, nodule_list, model_path="src/algorithms/classify/assets/g
 
     if torch.cuda.is_available():
         casenet = torch.nn.DataParallel(casenet).cuda()
-    # else:
+        # else:
```
Comment (WGierke) [inline prediction/src/algorithms/classify/src/gtr123_model.py]: I'm not sure if the indentation level is now correct?

Comment (vessemer) [reply prediction/src/algorithms/classify/src/gtr123_model.py]: Thanks :)

File: prediction/src/preprocess/crop_patches.py [hunk 2]
```diff
@@ -21,10 +21,12 @@ def mm2voxel(coord, origin=0., spacing=1.):
     origin = scipy.ndimage._ni_support._normalize_sequence(origin, len(coord))
     spacing = scipy.ndimage._ni_support._normalize_sequence(spacing, len(coord))
     coord = np.ceil(coord - np.array(origin)) / np.array(spacing)
+    print(coord)
```
Comment (lamby) [inline prediction/src/preprocess/crop_patches.py]: Committed by accident?

File: prediction/src/preprocess/preprocess_ct.py [hunk 3]
```diff
@@ -1,33 +1,40 @@
+import warnings
+
 import numpy as np
 import scipy.ndimage
 from src.preprocess import load_ct
 
 
 class Params:
-    """Params for CT data preprocessing.
+    """Params for CT data pre-processing.
```
Comment (lamby) [inline prediction/src/preprocess/preprocess_ct.py]: Please try and make these sorts of changes in a separate commit in future.. much easier to review  :)

File: prediction/src/preprocess/preprocess_ct.py [hunk 4]
```diff
@@ -52,61 +59,81 @@ def __init__(self, hu_transform=False, clip_lower=None, clip_upper=None,
             raise ValueError('The min_max_normalize should be bool or int')
         self.min_max_normalize = min_max_normalize
 
+        if not isinstance(scale, (int, float)) and (scale is not None):
+            raise ValueError('The scale should be float or int')
```
Comment (lamby) [inline prediction/src/preprocess/preprocess_ct.py]: ``TypeError`` would seem more suitable here.

File: prediction/src/preprocess/preprocess_ct.py [hunk 5]
```diff
@@ -52,61 +59,81 @@ def __init__(self, hu_transform=False, clip_lower=None, clip_upper=None,
...
+        self.scale = scale
+
+        if not isinstance(order, int) or not (4 >= order >= 0):
+            raise ValueError('The order should be int and lie in interval [0, 4]')
+        self.order = order
 
-class PreprocessCT:
+        if dtype not in np.typeDict.keys() and (dtype is not None):
+            raise ValueError('The dtype should be one of appropriate np.dtype')
```
Comment (lamby) [inline prediction/src/preprocess/preprocess_ct.py]: (ditto)

File: prediction/src/tests/test_load_ct.py [hunk 6]
```diff
@@ -68,11 +68,11 @@ def test_load_meta(metaimage_path, dicom_path):
 def test_metadata(metaimage_path, dicom_path):
     meta = load_ct.load_ct(dicom_path, voxel=False)
     meta = load_ct.MetaData(meta)
-    zipped = zip(meta.spacing, (0.703125, 0.703125, 2.5))
+    zipped = zip(meta.spacing, (2.5, 0.703125, 0.703125))
     assert all([m_axis == o_axis for m_axis, o_axis in zipped])
 
     meta = load_ct.load_ct(metaimage_path, voxel=False)
-    spacing = list(reversed(meta.GetSpacing()))
+    spacing = meta.GetSpacing()[::-1]
```
Comment (lamby) [inline prediction/src/tests/test_load_ct.py]: Hm? :)

File: prediction/src/preprocess/crop_patches.py (hunk 2 above)
Comment (vessemer) [reply prediction/src/preprocess/crop_patches.py]: Yep

File: prediction/src/preprocess/preprocess_ct.py (hunk 3 above)
Comment (vessemer) [reply prediction/src/preprocess/preprocess_ct.py]: Ok, thank you for the review :)

File: prediction/src/preprocess/preprocess_ct.py (hunk 4 above)
Comment (vessemer) [reply prediction/src/preprocess/preprocess_ct.py]: Indeed it is.

File: prediction/src/tests/test_load_ct.py (hunk 6 above)
Comment (vessemer) [reply prediction/src/tests/test_load_ct.py]: The `SimpleITK` reads CT scan array from MetaImages in `zyx` order, methods like GetSpacing | GetOrigin return values in `xyz` order, though.

Comment (lamby) [reply prediction/src/tests/test_load_ct.py]: Can you add this as a comment in the code? :)

File: prediction/src/tests/test_grt123_preprocess.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    """ """
+
+    def __init__(self):
+        self.crop_size = [96, 96, 96]
+        self.scaleLim = [0.85, 1.15]
+        self.radiusLim = [6, 100]
+        self.stride = 4
+        self.filling_value = 160
+
+    def __call__(self, imgs, target):
```
Comment (lamby) [inline prediction/src/tests/test_grt123_preprocess.py]: Kinda gross to abuse dunderscore call. I'm assuming this code is from elsewhere?

Comment (vessemer) [reply prediction/src/tests/test_grt123_preprocess.py]: Yes, this code had been taken from the [edited version](https://github.com/concept-to-clinic/concept-to-clinic/commit/45db7be01d0ebcd4a91a071b4055eee4aac949cf#diff-af6ce34696667e4dea35064c3e319127R180) of the original [SimpleCrop](https://github.com/lfz/DSB2017/blob/76c6d60831a7d94b01e3774ed449acd78a525b78/training/classifier/data_classifier.py#L110) class. I've put unchanged functions into [`test_grt123_preprocess.py`](https://github.com/concept-to-clinic/concept-to-clinic/pull/170/files/389824f6042a5cbc4cb537c531ae40446ab6164a#diff-0ea94b8ddb8cba945d0d72ca74d13fd3) in order to ensure absolute interchangeability.

**Comment by lamby** [comment]
Thanks!

**Pull request merged**